### PR TITLE
Make main page completed numbers consistent

### PR DESCRIPTION
**Pull request by windymilla**
As reported in [Task 1883](https://www.pgdp.net/c/tasks.php?action=show&task_id=1883), the number of "titles preserved for the world" does not match the number of Gold texts.

Use a common function that only counts distinct PG ebook numbers for both cases, and use "group by" when listing the Gold texts on the `list_etexts` page.

Also, fix bug in number of pages used to display the Gold texts: if the number of texts per page was an exact divisor of the total number of texts, it miscalculated the number of pages needed by one, the last of which was blank. Testing note: on TEST, there are currently 15 projects, but if shown 5 per page, tries to use 4 pages. With these changes, there are only 14 projects, which can be shown 7 per page on 2 pages.

Questions:
1. Using "group by" means that it is random (or rather "indeterminate")  which of the projects that share the PG etext number will be shown - is this OK?
2. Is "group by" bad performance-wise? 

Sandbox at: https://www.pgdp.org/~windymilla/c.branch/completed-numbers

### CODE REVIEW

**Review by cpeel — COMMENTED** [review]
re: `GROUP BY` performance -- like many DB things it's complicated: https://dev.mysql.com/doc/refman/8.0/en/group-by-optimization.html

But here's the explain plan from PROD for some actual data.

Current query:
```sql
EXPLAIN SELECT projectid, authorsname, nameofwork, language, n_pages, modifieddate, postednum, formats, state
FROM projects LEFT OUTER JOIN pg_books ON projects.postednum=pg_books.etext_number
WHERE (state='proj_submit_pgposted') ORDER BY modifieddate desc LIMIT 20 OFFSET 0;
```

```
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------+
| id | select_type | table    | partitions | type   | possible_keys | key           | key_len | ref                              | rows  | filtered | Extra               |
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------+
|  1 | SIMPLE      | projects | NULL       | ref    | state_moddate | state_moddate | 203     | const                            | 40882 |   100.00 | Backward index scan |
|  1 | SIMPLE      | pg_books | NULL       | eq_ref | PRIMARY       | PRIMARY       | 4       | dproofreaders.projects.postednum |     1 |   100.00 | Using where         |
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------+
```


With `GROUP BY`:
```sql
-- we need to disable only_full_group_by, we do this in DPDatabase for each DB session
SET SESSION sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));
EXPLAIN SELECT projectid, authorsname, nameofwork, language, n_pages, modifieddate, postednum, formats, state
FROM projects LEFT OUTER JOIN pg_books ON projects.postednum=pg_books.etext_number
WHERE (state='proj_submit_pgposted') GROUP by postednum ORDER BY modifieddate desc LIMIT 20 OFFSET 0;
```

```
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------------------+
| id | select_type | table    | partitions | type   | possible_keys | key           | key_len | ref                              | rows  | filtered | Extra                           |
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------------------+
|  1 | SIMPLE      | projects | NULL       | ref    | state_moddate | state_moddate | 203     | const                            | 40882 |   100.00 | Using temporary; Using filesort |
|  1 | SIMPLE      | pg_books | NULL       | eq_ref | PRIMARY       | PRIMARY       | 4       | dproofreaders.projects.postednum |     1 |   100.00 | Using where                     |
+----+-------------+----------+------------+--------+---------------+---------------+---------+----------------------------------+-------+----------+---------------------------------+
```

So we'll be using a temporary file & file sort vs an index scan, so it's less efficient but I ran both queries on PROD and they took the same amount of time. Given where this function is used I think this is fine.

File: pinc/showstartexts.inc [hunk 1]
```diff
@@ -9,25 +10,26 @@ function showstartexts($etext_limit, $type)
     if ($type == "bronze") {
         $aria_label_attr = attr_safe(_("Bronze Star"));
         $state = SQL_CONDITION_BRONZE;
+        $total = mysqli_num_rows(DPDatabase::query("SELECT projectid FROM projects WHERE $state"));
```
[the remark this answers is not in the document]
Comment (windymilla) [reply pinc/showstartexts.inc]: Previous code used `mysqli_num_rows()` but good opportunity to improve now code is split for gold/silver/bronze- thanks.

File: pinc/theme.inc [hunk 2]
```diff
@@ -877,6 +851,22 @@ function show_completed_projects($terse = false)
 
 // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
+function total_completed_projects()
```
[the remark this answers is not in the document]
Comment (windymilla) [reply pinc/theme.inc]: Thanks - moved and changed includes

**Comment by windymilla** [comment]
@cpeel - In fixing silver/bronze case, it became necessary in `list_projects()` to use `mysqli_num_rows()` to count total number of gold/silver/bronze. The alternative would be for `list_projects()` to either receive even more arguments ($join clause, $join_clause?) containing snippets of sql from the calling routine, or for `list_projects()` itself to know about gold/silver/bronze and the different sql needed for the 3 cases. This is because gold requires a `JOIN` and `GROUP BY`, but the others don't, and using `COUNT(*)` doesn't do what we want if you have `GROUP BY`. If `list_projects()` knew about gold/silver/bronze, it could either call `total_completed_projects()` for gold, or use a `SELECT(*)` without `JOIN` and `GROUP BY` for the silver/bronze case. It seems like that would make the code more complicated, though could potentially speed up the display of silver/bronze texts. Since on PROD, the number of silver/bronze texts is far fewer than gold texts, I didn't feel it was worth it, but I'm willing to reorganise things if you have a better plan.

**Comment by windymilla** [comment]
Updated sandbox at: https://www.pgdp.org/~windymilla/c.branch/completed-numbers

**Review by srjfoo — APPROVED** [review]

File: pinc/showstartexts.inc (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/showstartexts.inc]: I should have looked closer - appreciate you improving what was there before!

**Pull request merged**